File: app/src/main/java/com/futhark/android/seevoice/controller/fragment/SpecificationListFragment.java
package com.futhark.android.seevoice.controller.fragment;

import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;

import com.futhark.android.seevoice.R;
import com.futhark.android.seevoice.base.BaseFragment;
import com.futhark.android.seevoice.base.BaseFragmentActivity;
import com.futhark.android.seevoice.controller.adapter.SpecificationListAdapter;
import com.futhark.android.seevoice.model.database.SeeVoiceSqliteDatabaseHelper;
import com.futhark.android.seevoice.model.database.TableVoiceSpecification;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * list specification
 * Created by liuhr on 30/05/2017.
 */

public class SpecificationListFragment extends BaseFragment {
    @BindView(R.id.fragment_empty_list_view)
    ListView specificationListView;

    private SpecificationListAdapter listAdapter;
    private Cursor cursor;
    private SQLiteDatabase database;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View fragment = inflater.inflate(R.layout.fragment_empty_list, container, false);
        ButterKnife.bind(this, fragment);
        return fragment;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        listAdapter = new SpecificationListAdapter(getActivity(), null, true);
        specificationListView.setAdapter(listAdapter);
        database = new SeeVoiceSqliteDatabaseHelper(getActivity()).getReadableDatabase();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_exercise_record_standard, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.menu_exercise_record_standard:
                gotoRecordSpecification();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        cursor = database.query(TableVoiceSpecification.VoiceSpecificationEntry.TABLE_NAME,
                TableVoiceSpecification.COLUMNS, null, null, null, null, null);
        if(cursor != null){
            cursor.moveToFirst();
            listAdapter.swapCursor(cursor);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if(cursor != null){
            cursor.close();
            cursor = null;
        }
    }

    private void gotoRecordSpecification(){
        Intent goToRecordActivityIntent = new Intent(getActivity(), BaseFragmentActivity.class);
        goToRecordActivityIntent.putExtra(BaseFragmentActivity.DUTY_FATE_FRAGMENT_INTENT, RecordingFragment.class.getName());
        getActivity().startActivity(goToRecordActivityIntent);
    }


}

File: app/src/main/java/com/futhark/android/seevoice/feature/standard/SpecificationListFragment.kt
package com.futhark.android.seevoice.feature.standard

import android.content.Intent
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.os.Bundle
import android.view.*
import android.widget.AdapterView
import android.widget.ListView
import com.futhark.android.seevoice.R
import com.futhark.android.seevoice.base.BaseFragment
import com.futhark.android.seevoice.base.BaseFragmentActivity
import com.futhark.android.seevoice.model.database.SeeVoiceSqliteDatabaseHelper
import com.futhark.android.seevoice.model.database.TableVoiceSpecification


/**
 * list specification
 * Created by liuhr on 30/05/2017.
 */

class SpecificationListFragment : BaseFragment() {
  private var specificationListView: ListView? = null

  private var listAdapter: SpecificationListAdapter? = null
  private var cursor: Cursor? = null
  private var database: SQLiteDatabase? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setHasOptionsMenu(true)
  }

  override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
//    Log.i(TAG, "will create specification list fragment")
    return inflater!!.inflate(R.layout.fragment_empty_list, container, false)
  }

  override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)
    specificationListView = view.findViewById(R.id.fragment_empty_list_view)
    listAdapter = SpecificationListAdapter(activity, null, true, true)
    specificationListView!!.adapter = listAdapter
    database = SeeVoiceSqliteDatabaseHelper(activity).readableDatabase
    val onItemLongClickListener = AdapterView.OnItemLongClickListener { adapterView, view, i, l ->
      val item = adapterView.getItemAtPosition(i)
      val voiceSpecificationEntry = TableVoiceSpecification.VoiceSpecificationEntry(cursor)
      deleteTheRecord(voiceSpecificationEntry)
      true
    }
    specificationListView!!.onItemLongClickListener = onItemLongClickListener
  }

  override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
    super.onCreateOptionsMenu(menu, inflater)
    inflater.inflate(R.menu.menu_exercise_record_standard, menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.menu_exercise_record_standard -> {
        gotoRecordSpecification()
        return true
      }
    }
    return super.onOptionsItemSelected(item)
  }

  override fun onResume() {
    super.onResume()
    resumeCursor()
  }

  override fun onPause() {
    super.onPause()
    closeCursor()
  }

  private fun closeCursor() {
    cursor?.close()
    cursor = null
  }

  private fun resumeCursor() {
    closeCursor()
    cursor = database!!.query(TableVoiceSpecification.VoiceSpecification.TABLE_NAME,
        TableVoiceSpecification.COLUMNS, null, null,
        null, null, null)
    if (cursor == null) return
    cursor?.moveToFirst()
    if (cursor!!.isAfterLast) {
      gotoRecordSpecification()
    } else {
      listAdapter!!.swapCursor(cursor)
    }

  }

  private fun deleteTheRecord(voice: TableVoiceSpecification.VoiceSpecificationEntry) {
    val id = voice.id
    database?.delete(TableVoiceSpecification.VoiceSpecification.TABLE_NAME,TableVoiceSpecification.VoiceSpecification._ID + "= ?", arrayOf(id.toString()))
    resumeCursor()
  }

  private fun gotoRecordSpecification() {
    val goToRecordActivityIntent = Intent(activity, BaseFragmentActivity::class.java)
    goToRecordActivityIntent.putExtra(BaseFragmentActivity.DUTY_FATE_FRAGMENT_INTENT, SpecificationRecordingFragment::class.java.name)
    activity.startActivity(goToRecordActivityIntent)
  }


}
